batch.py

```diff
@@ -0,0 +1,568 @@
+        & (hl.len(mt.alleles) == 2)  # remove hom-ref
+        & (mt.n_unsplit_alleles == 2)  # biallelic
+        & (hl.is_snp(mt.alleles[0], mt.alleles[1]))  # SNVs
+    )
+
+    # filter rare variants only (MAF < 5%)
+    mt = hl.variant_qc(mt)
+    mt = mt.filter_rows(
+        (mt.variant_qc.AF[1] < 0.05) & (mt.variant_qc.AF[1] > 0)
+        | (mt.variant_qc.AF[1] > 0.95) & (mt.variant_qc.AF[1] < 1)
+    )
+    mt_path = output_path("nonref_qced_rare_variants.mt", "tmp")
+    mt = mt.checkpoint(mt_path, overwrite=True)  # checkpoint
+    logging.info(f"Number of rare variants (freq<5%): {mt.count()[0]}")
+
+    return mt_path
+
+
+# endregion GET_RELEVANT_VARIANTS
+
+
+# region GET_GENE_SPECIFIC_VARIANTS
+
+
+def get_promoter_variants(
+    mt_path: str,  # checkpoint from function above
+    ht_path: str,
+    gene_file: str,
+    gene_name: str,
+    window_size: int,
+    # plink_output_prefix: str, # figure out how to deal with this
+):
+    """Subset hail matrix table
+
+    Input:
+    mt_path: path to already subsetted hail matrix table
+    gene_file: path to a tsv file with information on
+    the chrom, start and end location of genes (rows)
+
+    Output:
+    For retained variants, that are: 1) in promoter regions and
+    2) within 50kb up or down-stream of the gene body (or in the gene body itself)
+    (on top of all filters done above)
+    
+    returns:
+    - Path to plink prefix for genotype files (plink format: .bed, .bim, .fam)
+    - Path to hail table with variant (rows) stats for downstream analyses
+    """
+    # read hail matrix table object (pre-filtered)
+    init_batch()
+    mt = hl.read_matrix_table(mt_path)
+
+    # get relevant chromosome
+    gene_df = pd.read_csv(AnyPath(dataset_path(gene_file)), sep="\t", index_col=0)
+    chrom = gene_df[gene_df["gene_name"] == gene_name]["chr"]
+    # subset to chromosome
+    mt = mt.filter_rows(mt.locus.contig == ("chr" + chrom))
+
+    # subset to window
+    # get gene body position (start and end) and build interval
+    interval_start = int(gene_df[gene_df["gene_name"] == gene_name]["start"]) - int(
+        window_size
+    )
+    interval_end = int(gene_df[gene_df["gene_name"] == gene_name]["end"]) + int(
+        window_size
+    )
+    # clip to chromosome boundaries
+    left_boundary = max(1, interval_start)
+    right_boundary = min(
+        interval_end, hl.get_reference("GRCh38").lengths[f"chr{chrom}"]
+    )
+    # get gene-specific genomic interval
+    gene_interval = f"chr{chrom}:{left_boundary}-{right_boundary}"
+    logging.info(f"Interval considered: {gene_interval}")  # 'chr22:23219960-23348287'
+
+    # include variants up to {window size} up- and downstream
+    mt = hl.filter_intervals(
+        mt, [hl.parse_locus_interval(gene_interval, reference_genome="GRCh38")]
+    )
+    logging.info(f"Number of variants within interval: {mt.count()[0]}")
+    logging.info(f"Number of variants in window: {mt.count()[0]}")
```

Comment:
This is the classic Hail mistake 😁 

Every time you run `count`, this will run all the lines from 136-164 (but not save the result), just to obtain the number. 

Where possible, always run `mt.count` after checkpointing, so the operations don't have to be re-run